runtime/autoload/man.vim

```diff
@@ -49,48 +46,52 @@ function! man#open_page(count, count1, ...) abort
       " user explicitly set a count
       let sect = string(a:count)
     endif
-    let [sect, name] = s:verify_exists(sect, name)
+    let [sect, name, path] = s:verify_exists(sect, name)
   catch
     call s:error(v:exception)
     return
   endtry
   call s:push_tag()
   let bufname = 'man://'.name.(empty(sect)?'':'('.sect.')')
-  let found_man = s:find_man()
-  if getbufvar(bufname, 'manwidth') ==# s:manwidth()
-    if found_man
-      silent execute 'buf' bufnr(bufname)
-    else
-      execute 'split' bufname
+  if a:mods !~# 'tab' && s:find_man()
+    if s:manwidth() ==# getbufvar(bufname, 'manwidth')
+      silent execute 'buf' bufname
+      call man#set_window_local_options()
+      keepjumps 1
+      return
     endif
+    noautocmd execute 'edit' bufname
+  elseif s:manwidth() ==# getbufvar(bufname, 'manwidth')
+    execute a:mods 'split' bufname
+    call man#set_window_local_options()
     keepjumps 1
     return
-  endif
-  if found_man
-    noautocmd execute 'edit' bufname
   else
-    noautocmd execute 'split' bufname
+    noautocmd execute a:mods 'split' bufname
   endif
-  call s:read_page(sect, name)
+  call s:read_page(path)
 endfunction
 
 function! man#read_page(ref) abort
   try
     let [sect, name] = s:extract_sect_and_name_ref(a:ref)
-    let [sect, name] = s:verify_exists(sect, name)
+    " The third element is the path.
+    call s:read_page(s:verify_exists(sect, name)[2])
   catch
     call s:error(v:exception)
     return
   endtry
-  call s:read_page(sect, name)
 endfunction
 
-function! s:read_page(sect, name) abort
+function! s:read_page(path) abort
   setlocal modifiable
   setlocal noreadonly
   keepjumps %delete _
   let b:manwidth = s:manwidth()
-  silent execute 'read!env MANWIDTH='.b:manwidth s:man_cmd s:man_args(a:sect, a:name)
+  " Ensure Vim is not recursively invoked (man-db does this)
+  " by forcing man to use cat as the pager.
+  " More info here http://comments.gmane.org/gmane.editors.vim.devel/29085
+  silent execute 'read!env MANPAGER=cat MANWIDTH='.b:manwidth s:man_cmd a:path
```

Comment:
It's good practice to have a space after `:read`:

    read !env

because `:write` behaves differently based on that. `:read` doesn't, but... it's a good practice.